src/test/regress/expected/function_propagation.out

```diff
@@ -316,6 +316,616 @@ BEGIN
+---------------------------------------------------------------------
+ localhost |    57637 | t       | (function,"{function_propagation_schema,func_in_transaction_10}",{integer})
+ localhost |    57638 | t       | (function,"{function_propagation_schema,func_in_transaction_10}",{integer})
+(2 rows)
+
+-- If constraint depends on a non-distributed table it should error out
+BEGIN;
+CREATE TABLE local_table_for_const(id int);
+CREATE OR REPLACE FUNCTION func_in_transaction_11(param_1 int, param_2 local_table_for_const)
+RETURNS boolean
+LANGUAGE plpgsql AS
+$$
+BEGIN
+    return param_1 > 5;
+END;
+$$;
+CREATE TABLE table_to_prop_func_9(id int, col_1 int check (func_in_transaction_11(col_1, NULL::local_table_for_const)));
+-- It should error out since there is non-distributed table dependency exists
+SELECT create_distributed_table('table_to_prop_func_9', 'id');
+ERROR:  type function_propagation_schema.local_table_for_const does not exist
+CONTEXT:  while executing command on localhost:xxxxx
+COMMIT;
+-- Show that function as a part of generated always is supporte
+BEGIN;
+   CREATE OR REPLACE FUNCTION non_sense_func_for_generated_always()
+   RETURNS int
+   LANGUAGE plpgsql IMMUTABLE AS
+   $$
+   BEGIN
+       return 1;
+   END;
+   $$;
+    -- Functions shouldn't be propagated within transaction
+    SELECT pg_identify_object_as_address(classid, objid, objsubid) from citus.pg_dist_object where objid = 'function_propagation_schema.non_sense_func_for_generated_always'::regproc::oid;
+ pg_identify_object_as_address
+---------------------------------------------------------------------
+(0 rows)
+
+   CREATE TABLE people (
+   id int,
+    height_cm numeric,
+    height_in numeric GENERATED ALWAYS AS (height_cm / non_sense_func_for_generated_always()) STORED);
+    SELECT create_distributed_table('people', 'id');
+ create_distributed_table
+---------------------------------------------------------------------
+
+(1 row)
+
+     -- Show that function is distributed after distributing the table
+    SELECT pg_identify_object_as_address(classid, objid, objsubid) from citus.pg_dist_object where objid = 'function_propagation_schema.non_sense_func_for_generated_always'::regproc::oid;
+                           pg_identify_object_as_address
+---------------------------------------------------------------------
+ (function,"{function_propagation_schema,non_sense_func_for_generated_always}",{})
+(1 row)
+
+COMMIT;
+-- Show that we don't need to propagate function as a part of the rule
+BEGIN;
+CREATE OR REPLACE FUNCTION func_for_rule()
+RETURNS int
+LANGUAGE plpgsql STABLE AS
+$$
+BEGIN
+    return 4;
+END;
+$$;
+CREATE TABLE table_1_for_rule(id int, col_1 int);
+INSERT INTO table_1_for_rule VALUES(7,7);
+CREATE TABLE table_2_for_rule(id int, col_1 int);
+INSERT INTO table_2_for_rule VALUES(7,7);
+CREATE RULE rule_1 AS ON UPDATE TO table_1_for_rule DO ALSO UPDATE table_2_for_rule SET col_1 = col_1 * func_for_rule();
```

Comment:
I guess with MX the rule should also exists on the workers. Not related to here, just noting which is already listed here: https://github.com/citusdata/citus/issues/4812

Reply:
It seems rules show up as Normal dependency. Lets fix that and improve the test here:
```SQL
SELECT objid, pg_identify_object_as_address(refclassid, refobjid, refobjsubid),deptype from pg_depend where objid = 17001;
┌───────┬───────────────────────────────────────────────────────┬─────────┐
│ objid │ pg_identify_object_as_address │ deptype │
├───────┼───────────────────────────────────────────────────────┼─────────┤
│ 17001 │ (table,"{public,table_1_for_rule}",{}) │ a │
│ 17001 │ ("table column","{public,table_2_for_rule,col_1}",{}) │ n │
│ 17001 │ (table,"{public,table_1_for_rule}",{}) │ n │
│ 17001 │ (function,"{public,func_for_rule}",{}) │ n │
└───────┴───────────────────────────────────────────────────────┴─────────┘
(4 rows)
Time: 2.108 ms
```